text/0000-stl-types.md

```diff
@@ -0,0 +1,168 @@
+
+The crux of the issue is that `vector` and `string` allocate and manage their memory using RAII and provide no facility for releasing ownership of that memory.
+
+In the examples below we'll just use `std::vector`. There are three cases we need to consider:
+
+```c++
+namespace foo {
+std::vector<float> returns();
+void as_mut_param(std::vector<float>& vec);
+void as_const_param(const std::vector<float>& vec);
+}
+```
+
+# Possible solutions
+
+## 1. Two-phase copy
+
+This is a modifcation of the error-message method above. For returning, we make two functions, say `len` and `get`. `len` is the function that actually gets the vector, then stashes it in a thread_local and returns its size, giving the caller an opportunity to allocate memory of the required size which it can pass to `get`, which will copy from the stashed vector:
+
+```c++
+int foo_returns_len() {
+    thread_local _stash = foo::returns();
+    return _stash.size();
+}
+
+void foo_returns_get(float* buffer, int len) {
+    memcpy(buffer, _stash.data(), sizeof(float) * min(_stash.size(), len));
+    _stash = {};
+}
+```
+
+`const_param` just copies into a temporary vector:
+```c++
+void foo_as_const_param(float* buffer, int len) {
+    std::vector tmp{buffer, len};
+    foo::as_const_param(tmp);
+}
+```
+
+`mut_param` is trickier since we potentially need to both pass data and return it. Easiest is probably just not to support passing data at all, in which case it reduces to the same as `returns`.
+
+### Pros
+- It's guaranteed to work and is safe
+- Caller is responsible for all allocations
+
+### Cons
+- Slow
+- Slightly ugly API for the `returns` case. `len` actually getting the vector is surprising for callers.
+- Works for POD elements only (no vector<string>)
+
+## 2. Binding-side storage
+
+In this model we stash the vector in a global (thread_local) map, indexed by its data pointer so it can be found and deleted later:
+
+```c++
+thread_local std::map<void*, std::any> _storage;
+
+foo_returns(float** buffer, int* len) {
+    std::vector<float> vec = foo::returns();
+    *buffer = vec.data();
+    *len = vec.size();
+    _storage[vec.data()] = std::move(vec);
+}
+
+void foo_delete_storage(void* buffer) {
+    _storage.erase(buffer);
+}
+```
+
+`const_param` has to invoke copies, same as method 1 since we have no way of constructing a vector directly. Same goes for `mut_param`, and even though we might be tempted to reuse an existing buffer from a previous call to `foo_returns()`, that would be very bad in the case of the vector reallocating.
+
+### Pros
+- The `returns` case is nice
+
+### Cons
+- Still copies in the parameter case
+- Works for POD elements only (no vector<string>)
```

Comment:
Another con is that it's really hard to make this pattern have a low time overhead, for many reasons, including cache misses on traversing the heap allocated nodes in the map

Reply:
Yeah a flat_map or a vector would be better here, but this is just a sketch. Still your point is valid.